aten/src/ATen/native/mps/operations/Attention.mm

```diff
@@ -0,0 +1,111 @@
+                                                                  const Tensor& key,
+                                                                  const Tensor& value,
+                                                                  const std::optional<Tensor>& attn_mask,
+                                                                  double dropout_p,
+                                                                  bool is_causal,
+                                                                  const std::optional<Tensor>& dropout_mask,
+                                                                  std::optional<double> scale) {
+  using namespace mps;
+  struct CachedGraph : public MPSCachedGraph {
+    CachedGraph(MPSGraph* graph) : MPSCachedGraph(graph) {}
+    MPSGraphTensor* qTensor = nil;
+    MPSGraphTensor* kTensor = nil;
+    MPSGraphTensor* vTensor = nil;
+    MPSGraphTensor* maskTensor = nil;
+    MPSGraphTensor* outputTensor = nil;
+    MPSGraphTensor* attnTensor = nil;
+  };
+  int64_t batchSize = query.size(0);
+  int64_t num_head = query.size(1);
+  int64_t qSize = query.size(2);
+  int64_t headSize = query.size(3);
+  int64_t maxSeqLength = key.size(2);
+  auto out = at::empty({batchSize, num_head, qSize, headSize}, query.options());
+  auto attn = at::empty({batchSize, num_head, qSize, maxSeqLength}, query.options());
+  auto scale_factor = sdp::calculate_scale(query, scale).sqrt().as_float_unchecked();
+  @autoreleasepool {
+    auto mkey = __func__ + getTensorsStringKey({query, key, value}) + ":" + std::to_string(is_causal) + ":" + std::to_string(attn_mask.has_value());
+    auto cachedGraph = LookUpOrCreateCachedGraph<CachedGraph>(mkey, [&](auto mpsGraph, auto graph) {
+      auto qTensor = mpsGraphRankedPlaceHolder(mpsGraph, query);
+      auto kTensor = mpsGraphRankedPlaceHolder(mpsGraph, key);
+      auto vTensor = mpsGraphRankedPlaceHolder(mpsGraph, value);
+      auto kT = [mpsGraph transposeTensor:kTensor dimension:2 withDimension:3 name:nil];
+      auto scaleTensor = [mpsGraph constantWithScalar:scale_factor shape:getMPSShape({1}) dataType:MPSDataTypeFloat32];
+
+      // sqTensor = qTensor * scale_factor
+      auto castedQ = qTensor;
+      if ([qTensor dataType] != MPSDataTypeFloat32) {
+        castedQ = [mpsGraph castTensor:qTensor toType:MPSDataTypeFloat32 name:nil];
+      }
+      auto sqTensor = [mpsGraph multiplicationWithPrimaryTensor:castedQ secondaryTensor:scaleTensor name:nil];
+      if ([sqTensor dataType] != qTensor.dataType) {
+        sqTensor = [mpsGraph castTensor:sqTensor toType:qTensor.dataType name:nil];
+      }
+
+      // skT = kT * scale_factor
+      auto castedKT = kT;
+      if ([kT dataType] != MPSDataTypeFloat32) {
+        castedKT = [mpsGraph castTensor:kT toType:MPSDataTypeFloat32 name:nil];
+      }
+      auto skT = [mpsGraph multiplicationWithPrimaryTensor:castedKT secondaryTensor:scaleTensor name:nil];
```

Comment:
Why are you not doing `q @ k.T` and then multiply by scale

Reply:
done!